src/cockpit/transports.py

```diff
@@ -413,6 +413,11 @@ def kill(self) -> None:
 
     def _close(self) -> None:
         if self._process is not None:
+            if self._process.stdin is not None:
+                try:
+                    self._process.stdin.close()
+                except ValueError:
```

Comment:
Where did you observe this exception happening?

I tried the following:

```
>>> import subprocess
>>> x = subprocess.Popen(['true'], stdin=subprocess.PIPE)
>>> x.stdin
<_io.BufferedWriter name=4>
>>> x.stdin.close()
>>> x.stdin.close()
>>> x.stdin.close()
>>> x.stdin.close()
>>> x.stdin.close()
>>> x.stdin.close()
>>> x.stdin.close()
```

and it never complained.

Also: can you take this out of the `if self._process is not None` block and move it to the bottom of the function.  It's the logical analogue of closing the `pty_fd`.

Reply:
At some point I got a "file descriptor is already closed" error, but I don't remember exactly where any more. I can try a run without it. I suppose it's actually cleaner to set `stdin` to None after closing it.

> can you take this out of the if self._process is not None

Well, then we'd get an `AttributeError` if self._process actually is None, no? `self.pty` is independent from that.

Reply:
Also, my intention was to close stdin before sending sigterm, that feels like a more friendly order.

Reply:
If that's the argument, then probably we should do the closing of the pipe-based stdin or the pty-based one both before killing the process.

I really feel fairly strongly that one way or the other, these two things are moral equivalents and belong directly beside one another.  I agree with your argument that it should be before the kill, but then please just move the kill to the end of the function.

Reply:
Right, makes sense! I moved it.